### PR TITLE
Getting backend countries from react-native

### PR DESCRIPTION
**Pull request by chrmod**
pair programmed with @khaled-cliqz 

In this PR we change the moment of react native context initialization. Now it start with on GeckoApplication startup. This is important so we can request backend countries even without Search UI being setup.

### CODE REVIEW

File: cliqz/index.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
 import React from 'react';
-import {AppRegistry, StyleSheet, Text, View, DeviceEventEmitter, } from 'react-native';
+import {AppRegistry, StyleSheet, Text, View, DeviceEventEmitter, NativeModules } from 'react-native';
 import SearchUI from 'browser-core/build/modules/mobile-cards-vertical/SearchUI';
 import App from 'browser-core/build/modules/core/app';
 import { Provider as CliqzProvider } from 'browser-core/build/modules/mobile-cards/cliqz';
+import console from 'browser-core/build/modules/core/console';
```
Comment (giota-cliqz) [inline cliqz/index.js]: Please remove

File: cliqz/index.js [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 
 class Cliqz {
   constructor(app) {
+    this.app = app;
```
Comment (giota-cliqz) [inline cliqz/index.js]: Is this used somewhere?

File: cliqz/index.js [hunk 3]
```diff
@@ -19,11 +21,20 @@ class BrowserCoreApp extends React.Component {
...
+        if (typeof id !== 'undefined') {
+          NativeModules.Bridge.replyToAction(id, response);
+        }
+        return response;
+      });
+    });
+  }
+
   componentWillMount() {
     const app = new App();
```
Comment (giota-cliqz) [inline cliqz/index.js]: Maybe you want to initialize in the constructor and use `this.app` here?

File: cliqz/index.js [hunk 4]
```diff
@@ -1,11 +1,13 @@
 import React from 'react';
-import {AppRegistry, StyleSheet, Text, View, DeviceEventEmitter, } from 'react-native';
+import {AppRegistry, StyleSheet, Text, View, DeviceEventEmitter, NativeModules } from 'react-native';
 import SearchUI from 'browser-core/build/modules/mobile-cards-vertical/SearchUI';
```
Comment (giota-cliqz) [inline cliqz/index.js]: We should import both `mobile-cards` and `mobile-cards-vertical` and have an if condition in the render method, that reads a configuration and decides on the style it should render

File: cliqz/index.js [hunk 5]
```diff
@@ -19,11 +21,20 @@ class BrowserCoreApp extends React.Component {
...
+  }
+
   componentWillMount() {
     const app = new App();
-    let cliqz;
-    const start = Date.now();
-    const loadingPromise = app.start().then(async () => {
+    this.loadingPromise = app.start().then(async () => {
       await app.ready();
       cliqz = new Cliqz(app);
```
Comment (giota-cliqz) [inline cliqz/index.js]: missing `let` or `const`?

File: cliqz/index.js (hunk 3 above)
Comment (chrmod) [reply cliqz/index.js]: does not matter really

File: cliqz/index.js (hunk 2 above)
Comment (chrmod) [reply cliqz/index.js]: in `onAction`

File: cliqz/index.js (hunk 1 above)
Comment (chrmod) [reply cliqz/index.js]: :+1:

File: mozilla-release/mobile/android/base/java/com/cliqz/react/SearchBackground.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import android.app.Application;
+import android.util.SparseArray;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class SearchBackground implements ReactInstanceManager.ReactInstanceEventListener {
+    private static SearchBackground ourInstance;
+    public final ReactInstanceManager mReactInstanceManager;
+    private ReactContext mReactContext;
+    private SparseArray<EventCallback> callbacks = new SparseArray<>();
```
Comment (spacifici) [inline mozilla-release/mobile/android/base/java/com/cliqz/react/SearchBackground.java]: final

File: mozilla-release/mobile/android/base/java/com/cliqz/react/SearchBackground.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        if (callback != null) {
+            callback.sendSuccess(response);
+        }
+    }
+
+    public static void startSearch(String query) {
+        getInstance().callAction("search", "startSearch", new Object[] { query });
+    }
+
+
```
Comment (spacifici) [inline mozilla-release/mobile/android/base/java/com/cliqz/react/SearchBackground.java]: Empty line

File: mozilla-release/mobile/android/base/java/com/cliqz/react/SearchUI.java [hunk 8]
```diff
@@ -1,54 +1,43 @@
...
-import com.facebook.react.common.LifecycleState;
-import com.facebook.react.shell.MainReactPackage;
 
 import org.mozilla.gecko.R;
 import org.mozilla.gecko.preferences.PreferenceManager;
 import org.mozilla.gecko.util.AppBackgroundManager;
-import org.mozilla.geckoview.BuildConfig;
 
 /**
  * @author Khaled Tantawy
```
Comment (spacifici) [inline mozilla-release/mobile/android/base/java/com/cliqz/react/SearchUI.java]: Replace with Cliqz Copyright

File: mozilla-release/mobile/android/base/java/com/cliqz/react/SearchBackground.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+                .setInitialLifecycleState(LifecycleState.RESUMED)
+                .build();
+        mReactInstanceManager.addReactInstanceEventListener(this);
+        mReactInstanceManager.createReactContextInBackground();
+        migrate();
+    }
+
+    private void migrate() {
+        // Let Search handle search lanaguge
+        PrefsHelper.getPref(GeckoPreferences.PREFS_SEARCH_REGIONAL, new PrefsHelper.PrefHandlerBase() {
```
Comment (spacifici) [inline mozilla-release/mobile/android/base/java/com/cliqz/react/SearchBackground.java]: Do it only on upgrade please

**Pull request merged**